### PR TITLE
Using SSHKit prefixes to hook RVM

### PR DESCRIPTION
**Pull request by kirs**
Here is the first implementation using SSHKit prefixes: https://github.com/leehambley/sshkit/pull/45

### CODE REVIEW

File: lib/capistrano/tasks/rvm.rake [hunk 1]
```diff
@@ -63,11 +43,13 @@ namespace :rvm do
     end
   end
 
+  before 'deploy:starting', 'rbenv:hook'
+  before 'deploy:starting', 'rbenv:check'
```
Comment (Kriechi) [inline lib/capistrano/tasks/rvm.rake]: should probably be RVM - not rbenv?

Comment (kirs) [inline lib/capistrano/tasks/rvm.rake]: My fault. Fixed

**Comment by Kriechi** [comment]
Bundler bin is not mapped any more. How do we address this issue - in combination with command map?

I think it is not a good idea to specify an addition rvm-map-bin entry inside capistrano-bundler...
So we need a sane default here in capistrano-{rvm,rbenv}

**Comment by kirs** [comment]
> Bundler bin is not mapped any more. How do we address this issue - in combination with command map?

Now it's done inside bundler: https://github.com/capistrano/bundler/pull/16/files#diff-91990b1845e3608397d87393341f0544R31

If you mean that someone is using `execute :bundle, "exec some_task"`, there are 2 ways:

1) Declare this way as not recommended and tell developers to add `some_task` to `bundle_bins` instead.
2) Keep mapping bundle to avoid regressions

**Comment by Kriechi** [comment]
Hmm - I mean, if I try `execute(:bundle, 'install')` - will RVM be invoked?
Because `:bundle` is not in rvm-map-bin...

Yes, if someone wants to `bundle exec` something, they have to add it to bundle-bins. Thats fine.

**Comment by kirs** [comment]
You're right, I've fixed it.

**Comment by kirs** [comment]
@Kriechi it works for me on capistrano 3.1 from upstream, SSHKit 1.2.0 and capistrano-rvm from this branch.

I would appreciate if you try it too :wink:

File: lib/capistrano/tasks/rvm.rake [hunk 2]
```diff
@@ -55,19 +28,28 @@ namespace :rvm do
...
+    rvm_prefix = "#{fetch(:rvm_path)}/bin/rvm #{fetch(:rvm_ruby_version)} do"
+    fetch(:rvm_map_bins).each do |command|
+      SSHKit.config.command_map.prefix[command.to_sym].unshift(rvm_prefix)
     end
   end
+end
 
+Capistrano::DSL.stages.each do |stage|
+  after stage, 'rvm:hook'
+  after stage, 'rvm:check'
```
Comment (Kriechi) [inline lib/capistrano/tasks/rvm.rake:46]: with that in place, the section https://github.com/capistrano/rvm#custom-tasks-which-rely-on-rvmruby is not needed any more, or is it?

Comment (kirs) [inline lib/capistrano/tasks/rvm.rake:46]: Fixed, thanks!

**Comment by Kriechi** [comment]
:+1:  It works!
Good to merge!

**Comment by kirs** [comment]
Awesome. Your tried the whole rails/bundler/rvm workflow, right?

**Comment by Kriechi** [comment]
```
  gem 'capistrano', github: 'capistrano/capistrano'
  gem 'capistrano-rvm',     github: 'capistrano/rvm', branch: 'sshkit-prefixes'
  gem 'capistrano-bundler', github: 'capistrano/bundler', branch: 'sshkit-prefixes'
  gem 'capistrano-rails',   github: 'capistrano/rails'
```

and 

```
require 'capistrano/setup'
require 'capistrano/deploy'

require 'capistrano/rvm'
require 'capistrano/rails'

# Loads custom tasks from `lib/capistrano/tasks' if you have any defined.
Dir.glob('lib/capistrano/tasks/*.cap').each { |r| import r }
```

`deploy.rb` and stage-files basically just like the templates.

**Pull request merged**